components/tikv_util/src/memory.rs

```diff
@@ -65,3 +74,105 @@ impl HeapSize for RaftCmdRequest {
             + mem::size_of_val(&self.status_request)
     }
 }
+
+pub struct MemoryQuota {
+    capacity: AtomicUsize,
+    in_use: AtomicUsize,
+}
+
+impl MemoryQuota {
+    pub fn new(capacity: usize) -> MemoryQuota {
+        MemoryQuota {
+            capacity: AtomicUsize::new(capacity),
+            in_use: AtomicUsize::new(0),
+        }
+    }
+
+    pub fn in_use(&self) -> usize {
+        self.in_use.load(Ordering::Relaxed)
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.capacity.load(Ordering::Acquire)
+    }
+
+    pub fn set_capacity(&self, capacity: usize) {
+        self.capacity.store(capacity, Ordering::Release)
+    }
+
+    pub fn alloc(&self, bytes: usize) -> bool {
+        let mut in_use_bytes = self.in_use.load(Ordering::Relaxed);
+        let capacity = self.capacity.load(Ordering::Acquire);
+        loop {
+            if in_use_bytes + bytes > capacity {
+                return false;
+            }
+            let new_in_use_bytes = in_use_bytes + bytes;
+            match self.in_use.compare_exchange_weak(
+                in_use_bytes,
+                new_in_use_bytes,
+                Ordering::Acquire,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => return true,
+                Err(current) => in_use_bytes = current,
+            }
+        }
+    }
+
+    pub fn free(&self, bytes: usize) {
```

Comment:
Same as alloc. 
I think we don't need to achieve absolute accurate on memory usage. MB level accuracy is enough.